### ISSUE TITLE
DSub support

### ISSUE THREAD

**Issue by PsyZeus** [ampache/ampache#2240]
Not sure if that's an issue or a drop but d-sub stopped working under 4.0.4 and shows this error.
![Screenshot_20191211-142735](https://user-images.githubusercontent.com/26334546/70624841-cd310480-1c29-11ea-9b11-9c36f1b46b8b.jpg)
![Screenshot_20191211-142804](https://user-images.githubusercontent.com/26334546/70624853-d4f0a900-1c29-11ea-8213-c232eb97e081.jpg)

* Ampache version: 4.0.4
* Web server + version: Apache 2.4.38
* Server operating system: Debian 10.1
* Client operating system: Android 7.1.2
* APP: Dsub 5.4.4

**Comment by lachlan-00** [comment]
I use dsub every day, you'll need to give me info about what the error is.
What you get when you put the link in a browser and what the webserver is.

There seem to be issues arround nginx rewriting if you use that

**Comment by kuzi-moto** [comment]
It's important to fill in the bug report template as completely as possible so we have all the information needed. The server logs in this case are pretty helpful.

**Comment by PsyZeus** [comment]
> 
> 
> I use dsub every day, you'll need to give me info about what the error is.
> What you get when you put the link in a browser and what the webserver is.
> 
> There seem to be issues arround nginx rewriting if you use that

I wish i knew what the error is! With 3.9.0 was working just fine but with 4.0.4 it gives this error.
I rolled back to 3.0.9 and is still working so that's why i though it has to do with the ampache version.
I have no problem with the webserver it's just the app that can't connect for some reason. under 4.0.4


> 
> 
> It's important to fill in the bug report template as completely as possible so we have all the information needed. The server logs in this case are pretty helpful.

Yea sorry about that.
I just edited the first post.

**Comment by wagnered** [comment]
@PsyZeus,

Is that error from clicking on the "Test connection" button?  If not, what is the response when you click on that button?   If the responses is successful then you should enter the request you issued.directly in the the local browser using the

**Comment by wagnered** [comment]
Oops...  big fingers, little keyboard.

I wanted to let you know that the information to format the request for your browser can be found on the [Subsonic API](http://www.subsonic.org/pages/api.jsp) document site.

**Comment by cdutriez** [comment]
Hi, same symptoms here with 4.0.x

WebServer works fine both locally and trough internet.
Dsub can't connect to server both locally and externally.

I rolled back to 3.9.0 and DSub android app works fine again.

**Comment by lachlan-00** [comment]
it might be token auth in subsonic 1.13.
ampache 4 supports both auth methods so i would suggest looking up this issue.

dsub is pretty good at determining which version of the api you use and picking an auth method.
https://github.com/ampache/ampache/issues/2210

**Comment by PsyZeus** [comment]
> 
> 
> @PsyZeus,
> 
> Is that error from clicking on the "Test connection" button? If not, what is the response when you click on that button? If the responses is successful then you should enter the request you issued.directly in the the local browser using the

Yes, the second screenshot is what i got after the "Test connection" button.

 

> 
> 
> it might be token auth in subsonic 1.13.
> ampache 4 supports both auth methods so i would suggest looking up this issue.
> 
> dsub is pretty good at determining which version of the api you use and picking an auth method.
> #2210

So is there a work around to fix that? how can i choose what method ampache use?

**Comment by wagnered** [comment]
> how can i choose what method ampache use?

Ampache does it automatically. 

Does the ampache log report any errors?  Does the webserver?  If its a problem with authentication, Ampache should report an error if debug is enabled.





-------- Original message --------
Subject: Re: [ampache/ampache] DSub support (#2240)

**Comment by cdutriez** [comment]
On my side : 
- connection test from different subsonic app (subsonic and dsub) don't work. Nothing appear in the logs (both apache2 and ampache)
- however, when using Dsub and trying to connect, it show at first no error. When trying to list songs,  I get a message saying that ressources are not found and that I should verify the server address. Meanwhile, I get theses logs on apache : 

> 80.215.81.75 - - [14/Dec/2019:20:47:43 +0000] "GET /rest/getMusicDirectory.view?u=admin&p=enc:6b6e6e6468786c31696d63696e796b77&v=1.2.0&c=DSub&id=100000261 HTTP/1.1" 404 3776 "-" "DSub"
> 80.215.81.75 - - [14/Dec/2019:20:48:10 +0000] "GET /rest/getMusicDirectory.view?u=admin&p=enc:6b6e6e6468786c31696d63696e796b77&v=1.2.0&c=DSub&id=100000285 HTTP/1.1" 404 3776 "-" "DSub"
>

**Comment by wagnered** [comment]
I noticed that your DSub client is using an old version of the subsonic API (1.2.0).  The latest is 1.13.0.

Could you insert the following rest URL into the browser on your server? localhost=127.0.0.1:

`HTTP://localhost/rest/getMusicDirectory.view?u=admin&p=enc:6b6e6e6468786c31696d63696e796b77&v=1.2.0&c=DSub&id=100000261`

If you get a well formed XML response, the problem might be with rewrite. If I remember correctly the "beautiful URL" setting must be enabled.

-------- Original message --------
Subject: Re: [ampache/ampache] DSub support (#2240)

**Comment by PsyZeus** [comment]
@cdutriez 
Exactly the same here to everything you wrote.

@wagnered 
Yes, ampache-access.log shows 1.2.0 for me too.

**Comment by lachlan-00** [comment]
Dsub reports 1.2.0 because that's the lowest API version it supports.

Pasting the URLs into Firefox and showing us what response you get would help

**Comment by PsyZeus** [comment]
@cdutriez
Exactly the same here to everything you wrote.

@wagnered
Yes, ampache-access.log shows 1.2.0 for me too.
What i got from the url you posted


under 3.0.9 is this
![h](https://user-images.githubusercontent.com/26334546/70859596-06fe5580-1f1f-11ea-9a1f-ba89e8000147.PNG)


under 4.0.4 is this
![c](https://user-images.githubusercontent.com/26334546/70859605-36ad5d80-1f1f-11ea-8a37-e4e082252206.PNG)

**Comment by cdutriez** [comment]
Same here, locally, but not on the server itself (no X on my raspbian) 👍  

3.9.0
```
<subsonic-response status="ok" version="1.11.0">
<directory id="100000261" name="Aerosmith">
<child id="...Hard Rock"/>
</directory>
</subsonic-response>
```

4.0.4
```Not Found

The requested URL was not found on this server.
Apache/2.4.38 (Raspbian) Server at ampache.dutriez.eu Port 443

**Comment by lachlan-00** [comment]
I just checked out master from develop to test and couldn't recreate. 404 seems like an odd error to get as well.

Does trying to bypass the rewrite rule work?
```
/rest/index.php?ssaction=getMusicDirectory&u=admin&p=enc:abcabcabc&v=1.2.0&c=DSub&id=100000285
```

**Comment by cdutriez** [comment]
I just typed this : 
`https://ampache.myserver.com//rest/index.php?ssaction=getMusicDirectory&u=admin&p=enc:abcabcabc&v=1.2.0&c=DSub&id=100000285`

and got that : 
`https://ampache.dutriez.eu//rest/index.php?ssaction=getMusicDirectory&u=admin&p=enc:abcabcabc&v=1.2.0&c=DSub&id=100000285`

**Comment by PsyZeus** [comment]
same here
![PNG](https://user-images.githubusercontent.com/26334546/70861090-513d0200-1f32-11ea-8e6e-844541bf6fda.PNG)

**Comment by lachlan-00** [comment]
If you get 404 from the subsonic urls and you get a correct response going to the ampache rule that means your rewrite rules aren't working

**Comment by lachlan-00** [comment]
Do a find in your web folder.
```
find ./ -name .htaccess
./rest/.htaccess
./server/.htaccess
./channel/.htaccess
./daap/.htaccess
./config/.htaccess
./bin/.htaccess
./play/.htaccess
```

**Comment by cdutriez** [comment]
done, it's ok now...

**Comment by PsyZeus** [comment]
In 4.0.4 rest, channel & play .htaccess are missing, i only have the dist file.
3.0.9 folder tho had created all of them when installed

**Comment by PsyZeus** [comment]
@cdutriez 
Do you mean you created the files and its working?

**Comment by lachlan-00** [comment]
That's your answer then! Ampache has to convert the subsonic request into on that it can use. :)

**Comment by cdutriez** [comment]
> 
> 
> @cdutriez
> Do you mean you created the files and its working?

You just copy the 3 files : 
```
cp /var/www/ampache/rest/.htaccess.dist /var/www/ampache/rest/.htaccess
cp /var/www/ampache/play/.htaccess.dist /var/www/ampache/play/.htaccess
cp /var/www/ampache/channel/.htaccess.dist /var/www/ampache/channel/.htaccess

```

**Comment by lachlan-00** [comment]
I think I should put a test in there for Apache to check these

**Comment by PsyZeus** [comment]
It's working! :D 
You all rock!!!
Anyway where is it mentioned what @cdutriez  mention because i followed the steps from upgrade wiki.

**Comment by cdutriez** [comment]
> 
> 
> It's working! :D
> You all rock!!!
> Anyway where is it mentioned what @cdutriez mention because i followed the steps from upgrade wiki.

https://github.com/ampache/ampache >> section **upgrading**

**Comment by PsyZeus** [comment]
Lol wtf?!! :P

**Comment by PsyZeus** [comment]
I guess it would be coll if someone put it on https://github.com/ampache/ampache/wiki/Upgrade

**Comment by cdutriez** [comment]
> 
> 
> I guess it would be coll if someone put it on https://github.com/ampache/ampache/wiki/Upgrade

Yes, it's missing in the documentation however : https://github.com/ampache/ampache/wiki/Upgrade

**Comment by PsyZeus** [comment]
Yep
Anyway that's why we are here! to help make it even better and useful for all.
Thank you all, well time music now! wooooo :)

**Comment by lachlan-00** [comment]
I've put a note in the wiki page but I'm also going to put a test in for this that can at least warn the user in logs

**Comment by wagnered** [comment]
@lachlan, while I was busy typing the following,  I noticed you solved their problem. Nice!

BTW

According to the Subsonic API specs, the 'v' parameter is:

"The protocol version implemented by the client, i.e., the version of the subsonic-rest-api.xsd schema used."

If DSub is doing what you claim then it isn't following the Subsonic specs.



-------- Original message --------
Subject: Re: [ampache/ampache] DSub support (#2240)

**Comment by lachlan-00** [comment]
Interesting. I asked this before on dsub's issues page but haven't really examined further

https://github.com/daneren2005/Subsonic/issues/967

**Comment by wagnered** [comment]
I read his comment.  It seems a bit convoluted.  I don't understand why he designed DSub contrary to his own specs.  

The following is according to my understanding of the subsonic API specs:

Ampache rest API is set at 1.13.0 which was created for Subsonic version 5.3.0 (October 2015). Many methods have been added since then.  The Subsonic API is now at 1.16.1. Can Ampache respond to methods added in 1.14.1 and later?  Theoretically Ampache should respond to DSub with error 20 (client must upgrade)  in response to all requests not implemented in version 1.2.0.

Changelog http://www.subsonic.org/pages/changelog.jsp

When you get some time, could you check out the Subsonic client and determine which version it displays: "minimum version", or actually used.

-------- Original message --------
Subject: Re: [ampache/ampache] DSub support (#2240)

**Comment by calcmandan** [comment]
I'm having the same problem. If you need any other information to troubleshoot this problem I would be happy to provide it.

Fresh ubuntu server, fresh ampache install. I get the same error.

**Comment by cdutriez** [comment]
If you read this thread, you'll see there is no more trouble.

**Comment by wagnered** [comment]
@calcmandan,

If your web server is Apache, copy the files in the following folders:

/rest/.htaccess.dist => /rest/.htaccess
/play/.htaccess.dist => /play/.htaccess
/channel/.htaccess.dist => /channel/.htaccess

If your server is Nginx, then go to https://github.com/ampache/ampache/wiki/Installation#Nginx.



-------- Original message --------
Subject: Re: [ampache/ampache] DSub support (#2240)

**Comment by calcmandan** [comment]
> If you read this thread, you'll see there is no more trouble.

I did read this thread, and it still didn't fix it for me. And I did as wagnered suggested well before I posted on here. The trick to making it work was changing the ownership of the .htaccess files from root to www-data. That's what fixed it. Surprised as I was, as I changed the ownership of the entire directory when I first copied it over to the /var/www/ directory. I suppose that doesn't cover hidden files.
